### DISCUSSION TITLE
I wonder if there is a Dockerfile, or Singularity build for GeNN?

### DISCUSSION THREAD

**Discussion by russelljjarvis** [genn-team/genn#545] [GeNN questions]
I wonder if there is a Dockerfile, or Singularity build for GeNN+CUDA tools? I am wondering this as I am have been debugging my GeNN installation regular GeNN works with CUDA but PyGeNN cannot build a CUDA backend, and so it doesn't have one.

Even if I succeed at debugging my laptop installation of PyGeNN I will ultimately have to redo the installation on HPC, and without administrative privileges.

I have noticed that a lot of HPC projects use the singularity to do simplify and containerize complex builds on HPC. Does something like this already exist for GeNN+PyGeNN?

**Comment by neworderofjamie** [comment]
What OS is your laptop running and how does PyGeNN fail? If C++ GeNN works ok, it shouldn't be too tricky to install - definitely no administrative access is required and, between us, we've installed it on several HPC systems without too much pain.

Creating a Dockerfile for GeNN is on my todo list but you would be more than welcome to help me out with it! The idea would be to have a docker image with an entrypoint that supports the kind of functionality of NEST's to launch interactive sessions and notebook environments (https://github.com/nest/nest-docker/blob/master/src/3.3/entrypoint.sh#L22-L50).

**Reply by russelljjarvis** [reply]
I am running PyGenn in Ubuntu+WSL, which may have some quirks. For some reason `~/.bashrc` is not sourcing automatically unless I manually source it when starting a terminal. Also, I neglected to read the readme.md inside the PyGenn directory, and I installed it incorrectly as a consequence, but both issues are fixed now.

I just made a Dockerfile for lava, so I can help start one up for GeNN too.

I think the idea is it would be CUDA+GeNN+PyGeNN+Sonata+common ecosystem tools that are taken for granted.

I think it should be layered Docker stacks where layer 1 is CUDA+GeNN+PyGeNN on Conda or minimal Ubuntu base, and layer 2 is the ecosystem tools you will need to build and analyze complex networks.

In a 3rd layer I would include Julia, as I want to benchmark GeNN against itself for the Summer of Code v1 model using regular LIF, versus GLIF cells, and then benchmark GeNN against Julia on the same model using this Julia+GPU approach: https://github.com/SpikingNetwork/TrainSpikingNet.jl

**Reply by russelljjarvis** [reply]
I kind of think for Windows users Docker might be better than WSL2. Another issue with WSL2 is it doesn't have access to all the system RAM by default.

**Reply by neworderofjamie** [reply]
I have successfully got WSL working on Windows 11 with CUDA interop but not so far on Windows 10 (https://docs.nvidia.com/cuda/wsl-user-guide/index.html were what I followed). Can you now build Windows docker images? If so, that might well be better and would provide a more familiar shell for Windows users

**Reply by russelljjarvis** [reply]
> Creating a Dockerfile for GeNN is on my todo list but you would be more than welcome **to help me out with it! The idea would be to have a docker image with an entrypoint that supports the kind of functionality of NEST's to launch interactive sessions and notebook environments (https://github.com/nest/nest-docker/blob/master/src/3.3/entrypoint.sh#L22-L50).**


Because of multistage builds, I think you could do:
```
FROM docker-registry.ebrains.eu/nest/nest-simulator:3.3
```
Somewhere near the end of the Dockerfile code I provided. That would cause some bloat, another approach would be to install 
```
FROM jupyter:scipynotebook # not a precise image name, just a gist.
```

**Reply by russelljjarvis** [reply]
> I have successfully got WSL working on Windows 11 with CUDA interop but not so far on Windows 10 (https://docs.nvidia.com/cuda/wsl-user-guide/index.html were what I followed). Can you now build Windows docker images? If so, that might well be better and would provide a more familiar shell for Windows users

Yes you can build windows Dockerimages, but its not something I have ever thought to do.
https://hub.docker.com/_/microsoft-windows-base-os-images

**Reply by russelljjarvis** [reply]
> > I have successfully got WSL working on Windows 11 with CUDA interop but not so far on Windows 10 (https://docs.nvidia.com/cuda/wsl-user-guide/index.html were what I followed). Can you now build Windows docker images? If so, that might well be better and would provide a more familiar shell for Windows users

-- Oh thanks for this you reminded me to check my host CUDA, I think I destroyed the host cudatools environment recently. 

> Yes you can build windows Dockerimages, but it's not something I have ever thought to do. https://hub.docker.com/_/microsoft-windows-base-os-images

For the idea of a based on the NEST container where there is an Jupyter notebook entry point, in that case, it might not matter that much what the Docker OS is.

I often like to do things like !pwd; !ls; or !wget inside notebooks, so I always naturally gravitate to Linux. Apparently powershell is POSIX compliant now, so occasionally GNU core util commands work in powershell.

**Comment by russelljjarvis** [comment]
I have made and tested out this Dockerfile
```dockerfile
FROM nvidia/cuda:11.8.0-devel-ubuntu22.04
ENV CUDA_PATH=/usr/local/cuda-11.8 
# RUN cuda-memcheck ./has_cuda ; exit 0
# RUN nvcc; exit 0
RUN apt update; 
RUN apt install git build-essential lcov wget -y --no-install-recommends 
ENV CONDA_DIR /opt/conda
RUN wget --quiet https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda.sh && \
     /bin/bash ~/miniconda.sh -b -p /opt/conda

ENV PATH=$CONDA_DIR/bin:$PATH
WORKDIR $HOME/workdir
RUN $(which conda) install -c miniconda swig numpy -y
RUN git clone https://github.com/genn-team/genn.git
RUN mv genn GeNN
WORKDIR GeNN
ENV PATH=$PATH:$HOME/workdir/GeNN/bin
RUN make DYNAMIC=1 LIBRARY_DIRECTORY=`pwd`/pygenn/genn_wrapper/
RUN python setup.py develop
WORKDIR $HOME/workdir/GeNN/userproject/PotjansMicrocircuit_project
RUN make
RUN ./generate_run test 
WORKDIR $HOME/workdir/GeNN/
RUN bash tests/run_tests.sh; exit 0 # This would fail any way for reasons outside of my control.
ENTRYPOINT ["/bin/bash"]
```

It fails at the step
```RUN ./generate_run test ```
but I think it is close.

Maybe I need a different Docker command for building with access to system cuda?

**Reply by neworderofjamie** [reply]
Thank you for this - I am actually going to a hackathon this afternoon to try and get some GeNN collaborators so I might try and get some more hands on this

**Reply by russelljjarvis** [reply]
I wonder if you have participated in BrainHack before? Its a good hackathon with international and local components. It used to be more focused on fMRI/EEG, but electrical neuronal network modelling is creeping in. 

You can contribute a project by making an issue for the repo at your local chapter.
https://github.com/brainhackorg/global2022
This is my one:
https://github.com/brainhackorg/global2022/issues/17

**Comment by neworderofjamie** [comment]
The docker file in Genn's docker2 branch works for me. With the nvidia docker stuff installed, you just need to run the container with ``docker run --gpus all`` to access the GPUs - will tidy it up tomorrow

**Reply by russelljjarvis** [reply]
Sweet. It's minimal. Great I love it! https://github.com/genn-team/genn/blob/docker2/Dockerfile
I will run it in a bit, thanks!

**Reply by russelljjarvis** [reply]
Maybe it was actually conda that was stuffing everything up in my build.

**Comment by neworderofjamie** [comment]
As of the 4.8.0 release, there is now an official GeNN docker image built from https://github.com/genn-team/genn/blob/master/Dockerfile and with prebuilt versions published at https://hub.docker.com/repository/docker/gennteam/genn
